### PR TITLE
Add `asarray_chkfinite` API.

### PR DESCRIPTION
**Pull request by khushi-411**
Hi, Team!
**Aim of the PR:** To implement `asarray_chkfinite` API.

It follows https://github.com/cupy/cupy/issues/6078#issue-1051191869.

Looking forward to your viewpoints. Thanks!

cc: @kmaehashi (:

### CODE REVIEW

File: cupy/_manipulation/kind.py
Comment (toslunar) [inline cupy/_manipulation/kind.py:8]:
```suggestion
    and raises an error if the input contains NaNs or Infs.
```

File: cupy/_manipulation/kind.py
Comment (toslunar) [inline cupy/_manipulation/kind.py:13]: It seems 'A' and 'K' are also supported.
https://numpy.org/doc/stable/reference/generated/numpy.asarray_chkfinite.html

File: cupy/_manipulation/kind.py
Comment (toslunar) [inline cupy/_manipulation/kind.py:16]: Google style cannot name the returned value.
```suggestion
        cupy.ndarray: Array interpretation of `a`.
```

**Review by toslunar — CHANGES_REQUESTED** [review]

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:14]: Could you add another test with a non-finite-value array? I am not sure if bool/int/complex types make sense here, so worth a look.

File: cupy/_manipulation/kind.py
Comment (khushi-411) [reply cupy/_manipulation/kind.py:16]: Noted (:

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py:14]: Hi, @leofang! 
I added the tests for the non-finite-value array, and all the tests are passed in the CI. Are more tests required?
> I am not sure if bool/int/complex types make sense here, so worth a look.

So, why did you think that was worth checking?
I checked for `numpy.asarray_chkfinite`. The following were the outputs: (which are as expected)

#### For bool
```
In [12]: a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=bool)

In [13]: a
Out[13]: array([ True, False,  True])
```
#### For int
```
In [14]: a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=int)
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-14-d6d91c379cab> in <module>
----> 1 a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=int)

~/anaconda3/lib/python3.8/site-packages/numpy/lib/function_base.py in asarray_chkfinite(a, dtype, order)
    481 
    482     """
--> 483     a = asarray(a, dtype=dtype, order=order)
    484     if a.dtype.char in typecodes['AllFloat'] and not np.isfinite(a).all():
    485         raise ValueError(

~/anaconda3/lib/python3.8/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
     81 
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84 
     85 

OverflowError: cannot convert float infinity to integer
```
#### For float
```
In [15]: a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=float)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-15-15536ad3bf83> in <module>
----> 1 a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=float)

~/anaconda3/lib/python3.8/site-packages/numpy/lib/function_base.py in asarray_chkfinite(a, dtype, order)
    483     a = asarray(a, dtype=dtype, order=order)
    484     if a.dtype.char in typecodes['AllFloat'] and not np.isfinite(a).all():
--> 485         raise ValueError(
    486             "array must not contain infs or NaNs")
    487     return a

ValueError: array must not contain infs or NaNs
```
#### For Complex
```
a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=complex)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-16-14e1c1a2ac61> in <module>
----> 1 a = numpy.asarray_chkfinite([numpy.inf, 0., -numpy.inf], dtype=complex)

~/anaconda3/lib/python3.8/site-packages/numpy/lib/function_base.py in asarray_chkfinite(a, dtype, order)
    483     a = asarray(a, dtype=dtype, order=order)
    484     if a.dtype.char in typecodes['AllFloat'] and not np.isfinite(a).all():
--> 485         raise ValueError(
    486             "array must not contain infs or NaNs")
    487     return a

ValueError: array must not contain infs or NaNs
```

**Comment by khushi-411** [comment]
Hi, @leofang! Hi, @toslunar!
Thank you so much for reviewing up the PR! I made all the suggested changes. Please let me know if anything else is required. I would be happy to learn (:

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (oke-aditya) [inline tests/cupy_tests/manipulation_tests/test_kind.py:23]: I believe this is what you need @khushi-411 

```suggestion
        self.assertRaises(ValueError, cupy.asarray_chkfinite, a, dtype):
```

This will check that the expected ValueError is raised when you are passing an array with infinite values.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (oke-aditya) [inline tests/cupy_tests/manipulation_tests/test_kind.py:24]: I think this too will raise an error and you won't get anything in `a_cpu`. As nothing would be returned.
I'm not sure you need to test this or even write this.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (oke-aditya) [inline tests/cupy_tests/manipulation_tests/test_kind.py:25]: Hence from above comment checking the dtype wouldn't make sense.

**Review by oke-aditya — COMMENTED** [review]
Little help as you asked offline. 

To other reviewers: -
@khushi-411 pinged me offline and asked for a small look. I thought I could help a bit.
I haven't contributed to CuPy but have experience in PyTorch

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py:24]: Yes, `a_cpu` too will raise value error.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Thanks a lot, @oke-aditya, for taking a look (: It worked.
But I am not too sure if we need this test or not.

Comment (oke-aditya) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: I believe you need the test. It is recommended to test the valueErrors as this is an expected behivour.

Any case the maintainers will tell you :)

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [reply tests/cupy_tests/manipulation_tests/test_kind.py:14]: > and all the tests are passed in the CI

Just FYI: actual GPU tests are not yet run, the green CIs are just preliminary checks (formatting, compilation, etc).

> > I am not sure if bool/int/complex types make sense here, so worth a look.
> 
> So, why did you think that was worth checking? 

Because experience showed that these are often corner cases in newly added APIs that require special treatments. It'd be nice to take care of them in one shot.

> I checked for `numpy.asarray_chkfinite`. The following were the outputs: (which are as expected)

Thank you, @khushi-411. Looks like the current implementation should work fine. I'll make a few minor, style suggestions shortly.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:14]: There are testing utilities that can help compare returned NumPy & CuPy arrays, parametrize array orders, etc. It looks like this file was rarely touched so it didn't get updated to use those, but it's never too late to modernize it 🙂
```suggestion
    @testing.for_orders('CFAK')
    @testing.for_all_dtypes()
    @testing.numpy_cupy_array_equal()
    def test_asarray_chkfinite(self, xp, dtype, order):
```

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:18]: Using the testing helpers (see above) allows you to write a lot less boilerplate codes and increase test coverages (after the return a lot of things are checked).
```suggestion
        return xp.asarray_chkfinite(a, dtype=dtype, order=order)
```

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:22]: Ditto. Also, we are phasing out the usage of `unittest`, and using pytest is preferred:  
```suggestion
    @testing.for_orders('CFAK')
    @testing.for_all_dtypes()
    def test_asarray_chkfinite_non_finite_vals(self, dtype, order):
```

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:23]: Nitpick: Ideally we also want to include NaN, though it's probably shadowed by the first non-finite entry... 
```suggestion
        a = [-numpy.inf, 0., numpy.inf, numpy.nan]
```

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:24]: See https://github.com/cupy/cupy/pull/3256 for how to check exceptions.
```suggestion
        for xp in (numpy, cupy):
            with pytest.raises(ValueError):
                xp.asarray_chkfinite(a, dtype=dtype, order=order)
```

**Review by leofang — CHANGES_REQUESTED** [review]
As mentioned earlier, suggesting a few style changes to be in line with current practice.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Please see my review comments below.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py:14]: > actual GPU tests are not yet run

Oo, that's the reason (: 
> It'd be nice to take care of them in one shot.

Ok, I understood. I'll try to figure out those cases the next time.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: > t looks like this file was rarely touched so it didn't get updated to use those, but it's never too late to modernize it :slightly_smiling_face: 

I totally agree with you and it seems Interesting. Shall I update other parts too? 
In this file, I think the following test cases can be updated:
- [test_asfortranarray](https://github.com/cupy/cupy/blob/0757aa31a10785e6b728ad465724e97a83945b73/tests/cupy_tests/manipulation_tests/test_kind.py#L21): All asfortranarray tests.
- [test_require_C_and_F_flags](https://github.com/cupy/cupy/blob/0757aa31a10785e6b728ad465724e97a83945b73/tests/cupy_tests/manipulation_tests/test_kind.py#L88)

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Got it! It's cooler than the previous approach.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Sure, I changed it!

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: > See #3256 for how to check exceptions.

Thanks for sharing the informative PR. I read that thoroughly (: 
I was wondering if a bit of enhancement can be made in code. For example, here, arrays are created inside the `for-loops`; that is not needed as it is independent of `xp`.
```
def test_dim_mismatch_ellipsis2(self):
        for xp in (numpy, cupy):
            a = xp.arange(12).reshape(2, 3, 2)
            with pytest.raises(ValueError):
                xp.einsum('i...,...i', a, a)
```
We can modify it to:
```
def test_dim_mismatch_ellipsis2(self):
        a = xp.arange(12).reshape(2, 3, 2)
        for xp in (numpy, cupy):
            with pytest.raises(ValueError):
                xp.einsum('i...,...i', a, a)
```
#### Gains
- Will take lesser computation time.

What do you think? Thanks for reading it (:

File: cupy/_manipulation/kind.py
Comment (leofang) [inline cupy/_manipulation/kind.py:8]: CuPy's convention is to start the short description right after the triple quotes. Also, correct the verb to be consistent:
```suggestion
    """Converts the given input to an array,
    and raises an error if the input contains NaNs or Infs.
```

File: cupy/_manipulation/kind.py
Comment (leofang) [inline cupy/_manipulation/kind.py:16]:
```suggestion
        cupy.ndarray: An array on the current device.
```

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:22]: We don't use the decorator `numpy_cupy_array_equal()` here, so `xp` is not needed:
```suggestion
    def test_asarray_chkfinite_non_finite_vals(self, dtype, order):
```

**Review by leofang — CHANGES_REQUESTED** [review]
Thanks, @khushi-411, just left a few minor suggestions. 

btw in case you don't know already: you can apply the suggested changes _at once_ easily via the GH web UI by going to the "Files changed" tab -> selecting "Add suggestions in batch" -> "Commit changes". Then, to check it locally you can just do `git pull` on the current branch.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Feel free to work on it! But I'd suggest to create a separate PR so that each PR's scope is well defined.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: No, `xp` is the loop variable, so it can't be pulled out of the loop.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Yeah, sure. Thanks!!

**Review by leofang — DISMISSED** [review]
LGTM! Thanks for contributing, @khushi-411.

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Oops, agreed. I was walking without looking into the other side :(

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (oke-aditya) [inline tests/cupy_tests/manipulation_tests/test_kind.py:25]: Hey Khushi @khushi-411 
I believe you can check the error message as well. 


```
with pytest.raises(ValueError, match='Expected error message'):
```

This is a good practice and ensures exact error is raised. You can use Regex etc but passing a string works well too.

Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py:25]: Hi, @oke-aditya!
Oo yes! Thanks for pointing out (: I'll keep a note!

Comment (leofang) [reply tests/cupy_tests/manipulation_tests/test_kind.py:25]: Hmmm I missed your [previous observation](https://github.com/cupy/cupy/pull/6275#discussion_r775804617) and so the CI failed, let's fix it...
```suggestion
            if xp.issubdtype(dtype, xp.integer):
                error = OverflowError
            else:
                error = ValueError
            with pytest.raises(error):
```

Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Here, we set `OverflowError` for all the dtypes lower/equal than integer type and the rest raises the `ValueError`.
I want to know more about the fixes you made and the error raised. The error message in the CI is:
```
=================================== FAILURES =================================== |  
-- | --
00:48:25.963796 STDOUT 2232] | _______________ TestKind.test_asarray_chkfinite_non_finite_vals ________________ |  
00:48:25.963800 STDOUT 2232] |   |  
00:48:25.963892 STDOUT 2232] | self = <cupy_tests.manipulation_tests.test_kind.TestKind testMethod=test_asarray_chkfinite_non_finite_vals> |  
00:48:25.963896 STDOUT 2232] | dtype = <class 'numpy.int8'>, order = 'C' |  
00:48:25.963902 STDOUT 2232] |   |  
00:48:25.963940 STDOUT 2232] | @testing.for_orders('CFAK') |  
00:48:25.963943 STDOUT 2232] | @testing.for_all_dtypes() |  
00:48:25.963950 STDOUT 2232] | def test_asarray_chkfinite_non_finite_vals(self, dtype, order): |  
00:48:25.964010 STDOUT 2232] | a = [-numpy.inf, 0., numpy.inf, numpy.nan] |  
00:48:25.964015 STDOUT 2232] | for xp in (numpy, cupy): |  
00:48:25.964020 STDOUT 2232] | with pytest.raises(ValueError): |  
00:48:25.964056 STDOUT 2232] | >               xp.asarray_chkfinite(a, dtype=dtype, order=order) |  
00:48:25.964058 STDOUT 2232] |   |  
00:48:25.964065 STDOUT 2232] | a          = [-inf, 0.0, inf, nan] |  
00:48:25.964113 STDOUT 2232] | dtype      = <class 'numpy.int8'> |  
00:48:25.964118 STDOUT 2232] | order      = 'C' |  
00:48:25.964126 STDOUT 2232] | self       = <cupy_tests.manipulation_tests.test_kind.TestKind testMethod=test_asarray_chkfinite_non_finite_vals> |  
00:48:25.964190 STDOUT 2232] | xp         = <module 'numpy' from '/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/numpy/__init__.py'>
```
Thank you!!

cc: @leofang

File: cupy/_manipulation/kind.py
Comment (kmaehashi) [inline cupy/_manipulation/kind.py:16]:
```suggestion
   .. note::
      This function performs device synchronization when ``a`` is a
      ``cupy.ndarray``.
```

File: cupy/_manipulation/kind.py
Comment (kmaehashi) [inline cupy/_manipulation/kind.py:21]: To avoid synchronization as much as possible, I think it's better to do inf check on CPU.

```py
if isinstance(a, cupy.ndarray):
    if not cupy.isfinite(a).all():
        raise ValueError(...)
else:
    a = cupy.asarray(np.asarray_chkfinite(a, dtype=dtype, order=order), dtype=dtype, order=order))
return a
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: In part of this test we convert infs/nans to ints. But they are not compatible and an `OverflowError` would be raised if `dtype` is of any integer type, so we need to branch out to let pytest catch different errors.

Looks like when `dtype` is bool nothing would be raised, so this test still failed. Let me suggest another approach shortly. (You can see why I said we need to check those corner cases 🙂)

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (leofang) [inline tests/cupy_tests/manipulation_tests/test_kind.py:21]: It does not make sense to test bool with nan/inf, so we simply disable it. (Alternatively, we can use `contextlib.nullcontext` for bool, which is available since Python 3.7, and a `pytest.raise` context manager for other dtypes. I don't have strong opinions here.) 
```suggestion
    @testing.for_all_dtypes(no_bool=True)
```

**Review by leofang — CHANGES_REQUESTED** [review]

File: tests/cupy_tests/manipulation_tests/test_kind.py
Comment (khushi-411) [reply tests/cupy_tests/manipulation_tests/test_kind.py]: Oo, that's the reason, Got It! Thanks!
> You can see why I said we need to check those corner cases 🙂

Yes, it's very much interesting 🙂

File: cupy/_manipulation/kind.py
Comment (leofang) [reply cupy/_manipulation/kind.py]: @kmaehashi Does it matter to avoid synchronization here if an exception is going to be raised anyway? Wouldn't it be better to raise sooner (by doing the check on GPU)? Put it differently, if the check passes (no exception) this takes longer than the original implementation.

Comment (kmaehashi) [reply cupy/_manipulation/kind.py]: Sorry my proposal was inaccurate, this is what I intended:
```py
if isinstance(a, cupy.ndarray):
    if not cupy.isfinite(a).all():
        raise ValueError(...)
else:
    a = np.asarray_chkfinite(a, dtype=dtype, order=order)
return cupy.asarray(a, dtype=dtype, order=order)
```

When an array is `cupy.ndarray`, there's no difference between the original implementation and my proposal. When an array is CPU array-like, the original implementation copies the array to GPU, checks for inf on GPU, and copies the result back to CPU. My proposal changes this to check for inf on CPU and (if no exceptions are raised) copy the array it to GPU.

Comment (khushi-411) [reply cupy/_manipulation/kind.py]: From the above discussion:
#### Original Implementation for CPU array
- Copies the arrays from CPU to GPU
- Perform Operations on GPU
- Copies back to CPU
#### Your Implementation for CPU array
- Performs operations on CPU
- Copies to GPU (if there are no infs and nans)
#### As per @leofang's comment
- If checks are passed, the implementation takes a longer time.
- GPU error's will be faster

@leofang, @kmaehashi, I have a small doubt. Like in the original implementation, we are making two implicit switches (CPU->GPU->CPU), while by your proposal, we are making a single switch (CPU->GPU). I am not too sure, but don't you think these *implicit* switches will affect the model's performance? Which factor to consider prior?

Thanks for the discussions! (:

Comment (leofang) [reply cupy/_manipulation/kind.py]: > When an array is `cupy.ndarray`, there's no difference between the original implementation and my proposal.

Yes.

> When an array is CPU array-like, the original implementation copies the array to GPU, checks for inf on GPU, and copies the result back to CPU. My proposal changes this to check for inf on CPU and (if no exceptions are raised) copy the array it to GPU.

Depends on our expectation for the likelihood of raising an error (due to either inf or nan) for this routine, the decision could differ. @kmaehashi has implicitly assumed that it's more likely to raise, whereas I expected in most cases the check will pass. In the former case, it's OK to do the check on host (which is slower) since the subsequent control flow will be returned to host, while in the latter case, it's cheaper to run the check on device (faster) and proceed to the next GPU calculation right away if there's no error.

Comment (kmaehashi) [reply cupy/_manipulation/kind.py]: Hmm I can be missing something.

> it's cheaper to run the check on device (faster) and proceed to the next GPU calculation right away

The evaluation of `if not cupy.isfinite(a).all()` always requires a copy of 0-dim bool array from device to host, so synchronization happens in any case (whether the inf check pass or not).

> I expected in most cases the check will pass

I also expect this :)

Comment (leofang) [reply cupy/_manipulation/kind.py]: Yes, but If the check runs much faster on GPU than on CPU, then it's probably ok to pay the sync cost 🙂

Comment (khushi-411) [reply cupy/_manipulation/kind.py:24]: cc: @kmaehashi @leofang

As per the above discussion, I updated the function considering speed as a priority.
Please let me know if anything else is required. I'll be happy to work on it.

File: cupy/_manipulation/kind.py
Comment (kmaehashi) [inline cupy/_manipulation/kind.py:18]:
```suggestion

    .. note::
        This function performs device synchronization.
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: cupy/_manipulation/kind.py
Comment (kmaehashi) [inline cupy/_manipulation/kind.py:16]:
```suggestion

    .. note::
```

File: cupy/_manipulation/kind.py
Comment (kmaehashi) [inline cupy/_manipulation/kind.py:26]:
```suggestion
            "array must not contain Infs or NaNs")
```

**Review by kmaehashi — APPROVED** [review]

**Comment by khushi-411** [comment]
Hi, @kmaehashi, @leofang, @toslunar, @oke-aditya!
Thank you so much for reviewing! :)

**Comment by kmaehashi** [comment]
@toslunar Could you approve if it's ok to you? 😄

**Review by leofang — APPROVED** [review]

**Comment by khushi-411** [comment]
Hi, @toslunar! A gentle ping for your approval :)

**Review by toslunar — APPROVED** [review]
LGTM

I'm sorry for keeping you waiting.

**Comment by leofang** [comment]
Thanks @khushi-411 and all!

**Pull request merged**
